### PR TITLE
Reconnect Message Flows When Collapsing Participant

### PR DESCRIPTION
**Pull request by philippfromme**
Reconnects message flows instead of removing them when a participant is collapsed.

nDTdeoPDrD

---

Related to https://github.com/camunda/camunda-modeler/issues/1651.

### CODE REVIEW

**Comment by barmac** [comment]
I've found a quirk when we don't have the straight layout:


https://user-images.githubusercontent.com/28307541/115020979-a63caf80-9ebb-11eb-89fd-695a4925add6.mov

**Review by barmac — CHANGES_REQUESTED** [review]
See the comment above.

**Comment by philippfromme** [comment]
@nikku I used the intermediate step to make sure the docking point is on the edge of the collapsed participant.

### With intermediate step:

SEwLykEyKO

The green dot marks the docking point for the first reconnect, the red dot marks the docking point for the second reconnect. Note that it is located at the edge.

### Without intermediate step:

5XGB8d9ifY

Without the intermediate step the red dot marks the docking point for the reconnect. It is located inside the participant as the end event was located inside it, too.

We could decide that this is okay and we don't need the intermediate step.

**Comment by nikku** [comment]
What is the benefit of (1)?

**Comment by barmac** [comment]
How do the solutions look like in the case of multi-edge connections?

**Comment by philippfromme** [comment]
@barmac 

With a first reconnect:

image

Without:

image

The only difference will be the `original` of the waypoints.

**Comment by barmac** [comment]
Both look OK 🍡

**Comment by philippfromme** [comment]
For simplicity, I will remove the first reconnect. I assume, you guys are okay with that?

**Comment by nikku** [comment]
> For simplicity, I will remove the first reconnect. I assume, you guys are okay with that?

Absolutely. Less computation and the result is visually exactly the same.

**Comment by philippfromme** [comment]
Ready to be reviewed again.

File: lib/features/modeling/behavior/RemoveParticipantBehavior.js [hunk 1]
```diff
@@ -41,8 +49,70 @@ export default function RemoveParticipantBehavior(eventBus, modeling) {
     }
   }, true);
 
+  this.postExecute('shape.replace', function(context) {
+    var oldShape = context.oldShape,
+        newShape = context.newShape;
+
+    if (!isCollapse(oldShape, newShape)) {
```
Comment (nikku) [inline lib/features/modeling/behavior/RemoveParticipantBehavior.js:56]: Better rename this to `isParticipantCollapse`?

File: lib/features/modeling/behavior/RemoveParticipantBehavior.js [hunk 2]
```diff
@@ -41,8 +49,70 @@ export default function RemoveParticipantBehavior(eventBus, modeling) {
     }
   }, true);
 
+  this.postExecute('shape.replace', function(context) {
```
Comment (nikku) [inline lib/features/modeling/behavior/RemoveParticipantBehavior.js:52]: Maybe add a comment what this part of the behavior is doing. Alternatively: Move the reconnect message flow behavior to an entirely different component. It is fairly standalone and I could imagine that I'll never again find where we implemented it in this file :see_no_evil:

**Review by nikku — CHANGES_REQUESTED** [review]
Comprehensive test coverage, I like!

After reviewing the code locally I was wondering if we could move it to an entirely different feature. That is just a thought though. Feel free to discard.

Works as specified. Small change, huge impact UX wise :clap:.

File: lib/features/modeling/behavior/RemoveParticipantBehavior.js (hunk 2 above)
Comment (philippfromme) [reply lib/features/modeling/behavior/RemoveParticipantBehavior.js:52]: Makes sense. There are limitations to the method of just adding new behaviors to an existing behavior that has an affected element in the file name. 🤡 I will move this to a separate behavior.

File: lib/features/modeling/behavior/RemoveParticipantBehavior.js (hunk 1 above)
Comment (philippfromme) [reply lib/features/modeling/behavior/RemoveParticipantBehavior.js:56]: 👍🏻

**Comment by barmac** [comment]
I couldn't break it. Great job!

**Review by barmac — APPROVED** [review]

File: lib/features/modeling/behavior/MessageFlowBehavior.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+import {
+  getResizedSourceAnchor,
+  getResizedTargetAnchor
+} from 'diagram-js/lib/features/modeling/cmd/helper/AnchorsHelper';
+
+/**
+ * BPMN-specific message flow behavior.
+ */
+export default function RemoveParticipantBehavior(eventBus, modeling) {
```
Comment (nikku) [inline lib/features/modeling/behavior/MessageFlowBehavior.js:19]: Can we rename it to `MessageFlowBehavior`, too? Also, in terms of what it does, why is it not called `ReconnectMessageFlowBehavior`?

I think we're 98.12% to finishing and merging this. I'm excited already :trumpet:

**Review by nikku — CHANGES_REQUESTED** [review]

File: lib/features/modeling/behavior/MessageFlowBehavior.js (hunk 3 above)
Comment (nikku) [reply lib/features/modeling/behavior/MessageFlowBehavior.js:19]: Looking into the final bits myself right now!

Comment (philippfromme) [reply lib/features/modeling/behavior/MessageFlowBehavior.js:19]: Ooops, sorry, I forgot to change the name.

Comment (philippfromme) [reply lib/features/modeling/behavior/MessageFlowBehavior.js]: Okay, done. 👍🏻

**Comment by philippfromme** [comment]
Oops, this was suddenly merged a split second after I pushed. 🤔

**Comment by nikku** [comment]
Good stuff we have it merged I guess :dancers:

**Pull request merged**